presto-main/src/main/java/io/prestosql/operator/GroupedTopNRankAccumulator.java

```diff
@@ -0,0 +1,894 @@
+        {
+            metricsBuffer.increment(groupId * 2);
+        }
+
+        public long getHeapSize(long groupId)
+        {
+            return metricsBuffer.get(groupId * 2 + 1);
+        }
+
+        public void setHeapSize(long groupId, long size)
+        {
+            metricsBuffer.set(groupId * 2 + 1, size);
+        }
+
+        public void addHeapSize(long groupId, long delta)
+        {
+            metricsBuffer.add(groupId * 2 + 1, delta);
+        }
+
+        public void incrementHeapSize(long groupId)
+        {
+            metricsBuffer.increment(groupId * 2 + 1);
+        }
+
+        public long sizeOf()
+        {
+            return INSTANCE_SIZE + heapIndexBuffer.sizeOf() + metricsBuffer.sizeOf();
+        }
+    }
+
+    /**
+     * Buffer abstracting storage of nodes in the heap. Nodes are referenced by their node index for operations.
+     */
+    private static class HeapNodeBuffer
+    {
+        private static final long INSTANCE_SIZE = ClassLayout.parseClass(HeapNodeBuffer.class).instanceSize();
+
+        /*
+         *  Memory layout:
+         *  [LONG] peerGroupIndex1, [LONG] peerGroupCount1, [LONG] leftChildNodeIndex1, [LONG] rightChildNodeIndex1,
+         *  [LONG] peerGroupIndex2, [LONG] peerGroupCount2, [LONG] leftChildNodeIndex2, [LONG] rightChildNodeIndex2,
+         *  ...
+         */
+        private final LongBigArray buffer = new LongBigArray();
+
+        private final LongBigArrayFIFOQueue emptySlots = new LongBigArrayFIFOQueue();
+
+        private long capacity;
+
+        /**
+         * Allocates storage for a new heap node.
+         *
+         * @return index referencing the node
+         */
+        public long allocateNewNode(long peerGroupIndex, long peerGroupCount)
+        {
+            long newHeapIndex;
+            if (!emptySlots.isEmpty()) {
+                newHeapIndex = emptySlots.dequeueLong();
+            }
+            else {
+                newHeapIndex = capacity;
+                capacity++;
+                buffer.ensureCapacity(capacity * 4);
+            }
+
+            setPeerGroupIndex(newHeapIndex, peerGroupIndex);
+            setPeerGroupCount(newHeapIndex, peerGroupCount);
+            setLeftChildHeapIndex(newHeapIndex, UNKNOWN_INDEX);
+            setRightChildHeapIndex(newHeapIndex, UNKNOWN_INDEX);
+
+            return newHeapIndex;
+        }
+
+        public void deallocate(long index)
+        {
+            emptySlots.enqueue(index);
+        }
+
+        public long getActiveNodeCount()
+        {
+            return capacity - emptySlots.longSize();
+        }
+
+        public long getPeerGroupIndex(long index)
+        {
+            return buffer.get(index * 4);
+        }
+
+        public void setPeerGroupIndex(long index, long peerGroupIndex)
+        {
+            buffer.set(index * 4, peerGroupIndex);
+        }
+
+        public long getPeerGroupCount(long index)
+        {
+            return buffer.get(index * 4 + 1);
```

Comment:
ditto. You could also extract `PEER_GROUP_COUNT_OFFSET, LEFT_CHILD_HEAP_INDEX_OFFSET`